source/common/common/utility.cc

```diff
@@ -58,34 +58,50 @@ std::string DateFormatter::fromTime(const SystemTime& time) const {
     cached_time.epoch_time_seconds = epoch_time_seconds;
   }
 
-  if (subsecond_widths_.empty()) {
+  if (subseconds_.empty()) {
     return cached_time.formatted.at(format_string_);
   }
 
-  // Copy the cached formatted string, then replace its subseconds part.
+  // Copy the current cached formatted format string, then replace its subseconds part.
   std::string formatted = cached_time.formatted.at(format_string_);
   const std::string value = fmt::FormatInt(epoch_time_ns.count()).str();
-  for (const auto width : subsecond_widths_) {
-    const std::string subsecond =
-        value.substr(SubsecondConstants::get().DEFAULT_SECONDS_LENGTH, width);
-    const size_t placeholder_index = formatted.find(SubsecondConstants::get().PLACEHOLDER_CHAR);
-    ASSERT(placeholder_index != std::string::npos);
-    formatted.replace(placeholder_index, width, subsecond);
+  for (const auto subsecond : subseconds_) {
+    // TODO(dio): Infer the length of second from parsing step. Currently, it is defaulted to 10.
+    const std::string digits = value.substr(10, subsecond.width_);
+    formatted.replace(subsecond.position_, subsecond.width_, digits);
   }
 
+  // TODO(dio): Assert the formatted length.
   return formatted;
 }
 
 std::string DateFormatter::parse(const std::string& format_string) {
+  auto now = std::chrono::system_clock::now();
+  time_t current_time = std::chrono::system_clock::to_time_t(now);
+  tm current_tm;
+  gmtime_r(&current_time, &current_tm);
+
   std::string new_format_string = format_string;
-  std::smatch result;
-  while (regex_search(new_format_string, result, SubsecondConstants::get().PATTERN)) {
-    const std::string& width_specifier = result[1];
+  std::smatch matched;
+  size_t position = 0;
+  size_t previous = 0;
+  std::array<char, 1024> buf;
+  while (regex_search(new_format_string, matched, SubsecondConstants::get().PATTERN)) {
+    const std::string& width_specifier = matched[1];
     const size_t width = width_specifier.empty() ? SubsecondConstants::get().PLACEHOLDER.size()
                                                  : width_specifier.at(0) - '0';
-    subsecond_widths_.emplace_back(width);
-    new_format_string.replace(result.position(), result.length(),
+    new_format_string.replace(matched.position(), matched.length(),
                               SubsecondConstants::get().PLACEHOLDER.substr(0, width));
+
+    const std::string part = new_format_string.substr(previous, matched.position() - previous);
+    const size_t formatted_length = strftime(&buf[0], buf.size(), part.c_str(), &current_tm);
```

Comment:
I think the part above now looks good, but I don't think you want to do anything that uses the current time representation here (it seems dangerous, since the formatted time at parse may be very different than the formatted time at request time). Here's my understanding:

1. When we parse, we are looking at the string without any substitutions, e.g. `%s-%3f-asdf-%9f`.
2. When we substitute, we replace the `%Nf` before doing the `strftime`. So, we don't need to do anything here that is post-`strftime`. 

The reason you need to do this additional complexity right now is due to how the cached time string works. I think what makes sense is to do the offset computation when you regenerate the cached item, in conjunctino with the `strftime`.

Sorry I missed this the first couple of rounds, that's the crux of it I think.

Reply:
@htuch I decided to have [`fromTimeAndPrepareSubsecondOffsets`](https://github.com/dio/envoy/blob/94e3addf09332b07d0b116329d66281672795611/source/common/common/utility.cc#L119-L152) to figure out the subsecond offsets of a format string after it is formatted. Hence I can use it [here](https://github.com/dio/envoy/blob/94e3addf09332b07d0b116329d66281672795611/source/common/common/utility.cc#L61-L65). 

I'm not super happy about it since it is a bit complex. Want to get your input on this. Thanks!